### PR TITLE
Limit repository size

### PR DESCRIPTION
**Pull request by sapk**
The goal of this PR is to define a repo limit size. I think org and user level restriction could come later.

#### TODO:
- [x] Calculate Push Size (I still have some corner-case to test mostly to not block deletion and some force push)
- [x] Edit max repo size
- [x] Enforce repo size
- [ ] Add tests
- [ ] ~Validate LFS~ LFS object are not calculated in .Size repo attr so I will skip LFS checks.

<details>
  <summary><b>Screenshots:</b></summary>

image
image
image
image
image
image

</details>

#### TOFIX :
<details>
  <summary>Deletion of file from UI trigger 500 when repo is over. -> TODO catch this specific error.</summary>
2019/08/16 05:23:58 ...uters/repo/editor.go:432:DeleteFilePost() [E] DeleteRepoFile: git push: remote: Gitea: new repo size is over limitation 10000        
	To /home/sapk/go/src/code.gitea.io/gitea/data/repositories/sapk/test.git
	 ! [remote rejected] d9629b41f9c58da756cf806aabf5811b1ff45b50 -> master (pre-receive hook declined)
	error: impossible de pousser des références vers '/home/sapk/go/src/code.gitea.io/gitea/data/repositories/sapk/test.git'
</details>

<details>
  <summary>Creation of branch from UI trigger 500 when repo is over. -> TODO catch this specific error.</summary>
2019/08/16 05:28:42 ...uters/repo/branch.go:287:CreateBranch() [E] CreateNewBranch: Push: exit status 1 - remote: Gitea: new repo size is over limitation 10000        
	To /home/sapk/go/src/code.gitea.io/gitea/data/repositories/sapk/test.git
	 ! [remote rejected] test -> test (pre-receive hook declined)
	error: impossible de pousser des références vers '/home/sapk/go/src/code.gitea.io/gitea/data/repositories/sapk/test.git'
</details>

Related: #3658

### CODE REVIEW

**Comment by oscarlofwenhamn** [comment]
Nice, this will be a good function. 👍

**Comment by sapk** [comment]
This PR is near ready it just miss the tests. You can give it a try.

File: models/repo.go [hunk 1]
```diff
@@ -675,6 +676,11 @@ func (repo *Repository) UpdateSize() error {
 	return repo.updateSize(x)
 }
 
+// RepoSizeIsOversized return if is over size limitation
+func (repo *Repository) RepoSizeIsOversized(additionnalSize int64) bool {
+	return repo.SizeLimit > 0 && repo.Size+additionnalSize > repo.SizeLimit
```
Comment (techknowlogick) [inline models/repo.go]: `additionnal`-> `additional`

File: modules/git/repo_tree.go [hunk 2]
```diff
@@ -95,3 +95,13 @@ func (repo *Repository) CommitTree(sig *Signature, tree *Tree, opts CommitTreeOp
 	}
 	return NewIDFromString(strings.TrimSpace(res))
 }
+
+//GetDiffTree Compute the diff-tree for sparse-checkout
+func GetDiffTree(repoPath, base, head string) (string, error) {
```
Comment (sapk) [inline modules/git/repo_tree.go]: This change is a non related refactor done during testing. Let me know if I should remove it.

**Comment by oscarlofwenhamn** [comment]
Is this dead?

**Comment by sapk** [comment]
@oscarlofwenhamn not really time go into it to fix the few nits. I will maybe split this PR in two: display (done), and enforce (need more work).

**Comment by oscarlofwenhamn** [comment]
@sapk Fully understandable and relatable. Splitting it into two would be brilliant, that way we could get the finished functionality going and maybe someone could more easily help out on the enforce-PR to iron out the last creases without risking the finished stuff.

**Comment by 6543** [comment]
@sapk ping

**Comment by sapk** [comment]
@6543 I worked first on splitting up a PR related to attachement security. I will do this one when the two last part of the first are merged.

**Comment by sapk** [comment]
Yes stale bot I should refresh this one. :smile:

**Comment by 6543** [comment]
do you think it will be ready for 1.12 :sweat_smile:

**Comment by sapk** [comment]
@6543 nope :smile:. It depend of free time that I don't have that much lately :sob:.

**Comment by oscarlofwenhamn** [comment]
Fingers crossed for this not dying. 🤞

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -250,6 +250,8 @@ var migrations = []Migration{
 	NewMigration("fix publisher ID for tag releases", fixPublisherIDforTagReleases),
 	// v157 -> v158
 	NewMigration("ensure repo topics are up-to-date", fixRepoTopics),
+	// v158 -> v157
```
Comment (nuno-silva) [inline models/migrations/migrations.go]: do you mean `v158 -> v159`?

Comment (sapk) [reply models/migrations/migrations.go]: Good catch

**Comment by sapk** [comment]
I rebased and cleaned this PR. It still need work and will try to finish it in the next days.

**Comment by lunny** [comment]
And should also include releases binaries?

**Comment by lafriks** [comment]
We should split lfs size to different column before this as lfs size is calculated into size now :(

**Comment by sapk** [comment]
@lunny and @lafriks, If you are ok with it I will keep this one on what is included in the current size calculation and when all the checks are in place we could improve them by splitting and adding more details on every parts.

**Comment by lunny** [comment]
CI failed and conflicted.

**Comment by DmitryFrolovTri** [comment]
@sapk Hi is this one still alive?

**Comment by DmitryFrolovTri** [comment]
Ok guys, I am planning to hire a developer to get this resolved hopefully I can get enough money, I estimate about 20h max on at 50$. Whatever remains would be allocated to another fix. Here is where I am trying to get the needed money:
https://opencollective.com/oss-code/contribute/support-gitea-limit-repository-size-7833-initiative-47842

**Comment by techknowlogick** [comment]
Thanks @DmitryFrolovTri, I emailed you, but I will also say here I'm having trouble backing this collective, but once resolved there will be some additional funds in that bounty.

cc: @kdumontnu who has supported bounties in the past who may also find this beneficial.

Should this bounty be funded I will also set aside some funds for @sapk for his work on this PR so far.

**Comment by kdumontnu** [comment]
> Thanks @DmitryFrolovTri, I emailed you, but I will also say here I'm having trouble backing this collective, but once resolved there will be some additional funds in that bounty.
> 
> cc: @kdumontnu who has supported bounties in the past who may also find this beneficial.
> 
> Should this bounty be funded I will also set aside some funds for @sapk for his work on this PR so far.

takemymoney

Yeah, I can probably come up with something.

**Comment by DmitryFrolovTri** [comment]
@kdumontnu well I think this conversation is not for this thread :) But my plan to use them on other gitea / opencollective there is also some admin costs which might be needed - OC keeps track of all this.

**Comment by DmitryFrolovTri** [comment]
@kdumontnu hi, seems some bank details were incorrect I fixed that. My plan (worst case) if we don't reach the goal donate back to gitea minus some processing/keeping costs

**Comment by sapk** [comment]
Feel free to take over. 
I don't really have time to finish this in the neartime and for the bounty you can do as you like. I would be already pleased with this feature being available in Gitea.

**Comment by DmitryFrolovTri** [comment]
Hi all, working on this here: 
https://github.com/DmitryFrolovTri/gitea/tree/limit-repo-size

**Comment by DmitryFrolovTri** [comment]
@sapk please give access to your PR to contribute or add as contributor to the sapk fork so I could push to limit-repo-size branch.

**Comment by techknowlogick** [comment]
closing in favour of https://github.com/go-gitea/gitea/pull/21820
